# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1459389217 25200
#      Wed Mar 30 18:53:37 2016 -0700
# Node ID feaf0b0e0b1f72225e94e0c3ef8e23c45ad1dcb3
# Parent  e2ed4c95d68806631bf8ed8eaf555678a3696175
test: filter fntests.log when running make check with V=1

* test/show-failures.awk: New file to filter fntests.log for test failures.
* test/module.mk (test_EXTRA_DIST): Include test/show-failures.awk in the list.

diff --git a/test/module.mk b/test/module.mk
--- a/test/module.mk
+++ b/test/module.mk
@@ -67,17 +67,17 @@ test/.gdbinit: etc/gdbinit
 
 define run-octave-tests
   cd test && $(SHELL) ../run-octave $(RUN_OCTAVE_OPTIONS) $(1) --norc --silent --no-history $(abs_top_srcdir)/test/fntests.m $(abs_top_srcdir)/test
   if $(AM_V_P); then \
     echo ""; \
     if [ -f test/fntests.log ]; then \
       echo "Contents of test/fntests.log:"; \
       echo ""; \
-      cat test/fntests.log; \
+      $(AWK) -f $(srcdir)/build-aux/show-failures.awk test/fntests.log; \
     else \
       echo "test/fntests.log is missing!"; \
     fi; \
   fi
 endef
 
 check-local: $(GENERATED_TEST_FILES) | $(OCTAVE_INTERPRETER_TARGETS) test/$(octave_dirstamp)
 	$(call run-octave-tests)
@@ -170,16 +170,17 @@ uninstall-test-files:
 BUILT_SOURCES += $(GENERATED_TEST_FILES)
 
 test_EXTRA_DIST += \
   test/build-conv-tests.sh \
   test/build-sparse-tests.sh \
   test/build-bc-overload-tests.sh \
   test/bc-overloads-expected \
   test/build_bc_overloads_expected.m \
+  test/show-failures.awk \
   $(TEST_FILES)
 
 EXTRA_DIST += $(test_EXTRA_DIST)
 
 test_CLEANFILES += \
   $(GENERATED_BC_OVERLOADS_FILES) \
   $(GENERATED_TEST_FILES)
 
diff --git a/test/show-failures.awk b/test/show-failures.awk
new file mode 100644
--- /dev/null
+++ b/test/show-failures.awk
@@ -0,0 +1,64 @@
+#! /usr/bin/awk -f
+#
+# Copyright (C) 2016 Mike Miller
+#
+# This file is part of Octave.
+#
+# Octave is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at
+# your option) any later version.
+#
+# Octave is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Octave; see the file COPYING.  If not, see
+# <http://www.gnu.org/licenses/>.
+
+BEGIN {
+  buf = "";
+  limit_output = 0;
+  line = 0;
+  line_limit = 32;
+}
+## Make the decision to print or not when we get to the next file
+/^>>>>> |^Files with no tests:/ {
+  print buf;
+
+  buf = "";
+  line = 0;
+  limit_output = 0;
+}
+## The first line of the output from a failing test
+/^\*\*\*\*\* / {
+  line = 0;
+  limit_output = 0;
+}
+/^!!!!! known (bug|failure)|^----- skipped test/ {
+  limit_output = 1;
+}
+## Accumulate all lines that display the test code and up to line_limit
+## lines of output from failed test.
+{
+  if (buf == "")
+    {
+      buf = $0;
+    }
+  else
+    {
+      if (limit_output)
+        {
+          if (line < line_limit)
+            buf = buf "\n" $0;
+          else if (line == line_limit)
+            buf = buf "\n[skipping remaining output]";
+
+          line++;
+        }
+      else
+        buf = buf "\n" $0;
+    }
+}
